### PR TITLE
Add workspace tagging for java

### PR DESCRIPTION
**Pull request by Flanker32**
This PR add more workspace tagging for java projects, including gradle, dependencies in pom/gradle and android manifest.

### CODE REVIEW

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 1]
```diff
@@ -468,10 +512,70 @@ export class WorkspaceTagsService implements IWorkspaceTagsService {
 					// Ignore errors when resolving file or parsing file contents
 				}
 			});
-			return Promise.all([...packageJsonPromises, ...requirementsTxtPromises, ...pipfilePromises]).then(() => tags);
+
+			const pomPromises = getFilePromises('pom.xml', this.fileService, this.textFileService, content => {
+				try {
+					let dependenciesContent;
+					while (dependenciesContent = dependenciesRegex.exec(content.value)) {
+						let dependencyContent;
```
Comment (rheh) [inline src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:520]: Isn't the following lines of code the same as calling processGradleDependencies?

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 2]
```diff
@@ -468,10 +512,70 @@ export class WorkspaceTagsService implements IWorkspaceTagsService {
...
+				this.tagJavaDependency(groupId, artifactId, 'workspace.gradle.', tags);
+			}
+		}
+	}
+
+	private tagJavaDependency(groupId: string, artifactId: string, prefix: string, tags: Tags): void {
+		for (const javaLibrary of JavaLibrariesToLookFor) {
+			if ((groupId === javaLibrary.groupId || new RegExp(javaLibrary.groupId).test(groupId)) &&
+				(artifactId === javaLibrary.artifactId || new RegExp(javaLibrary.artifactId).test(artifactId))) {
+				tags[prefix + javaLibrary.tag] = true;
```
Comment (rheh) [inline src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:574]: Can mutating tags be avoided by return the result?

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 3]
```diff
@@ -118,6 +118,49 @@ const PyModulesToLookFor = [
 	'playwright'
 ];
 
+let JavaLibrariesToLookFor: { groupId: string, artifactId: string, tag: string }[] = [
```
[the remark this answers is not in the document]
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Yes, fixed

[the remark this answers is not in the document]
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Moved to `javaWorkspaceTags`

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 4]
```diff
@@ -118,6 +118,49 @@ const PyModulesToLookFor = [
...
+	{ 'groupId': 'redis.clients', 'artifactId': 'jedis', 'tag': 'redis' },
+	{ 'groupId': 'org.redisson', 'artifactId': 'redisson', 'tag': 'redis' },
+	{ 'groupId': 'io.lettuce', 'artifactId': 'lettuce-core', 'tag': 'redis' },
+	// azure
+	{ 'groupId': 'com.microsoft.azure', 'artifactId': 'azure', 'tag': 'azure' },
+	{ 'groupId': 'com.microsoft.azure', 'artifactId': 'azure-mgmt-.*', 'tag': 'azure' },
+	{ 'groupId': 'com\\.microsoft\\.azure\\..*', 'artifactId': 'azure-mgmt-.*', 'tag': 'azure' },
+	// lombok
+	{ 'groupId': 'org.projectlombok', 'artifactId': 'lombok', 'tag': 'lombok' }
+];
```
[the remark this answers is not in the document]
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Change to A-Z order by tags

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 5]
```diff
@@ -118,6 +118,49 @@ const PyModulesToLookFor = [
...
+
+const dependenciesRegex = /<dependencies>([\s\S]*?)<\/dependencies>/g;
+const dependencyRegex = /<dependency>([\s\S]*?)<\/dependency>/g;
+const groupIdRegex = /<groupId>([\s\S]*?)<\/groupId>/;
+const artifactIdRegex = /<artifactId>([\s\S]*?)<\/artifactId>/;
+const gradleLooseRegex = /group\s*:\s*[\'\"](.*?)[\'\"]\s*,\s*name\s*:\s*[\'\"](.*?)[\'\"]\s*,\s*version\s*:\s*[\'\"](.*?)[\'\"]/g;
+const gradleCompactRegex = /[\'\"]([^\'\"\s]*?)\:([^\'\"\s]*?)\:([^\'\"\s]*?)[\'\"]/g;
```
[the remark this answers is not in the document]
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Done~

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 1 above)
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:478]: Maven differs a bit with gradle as I need to use different regex to get `groupId` and `artifactId` for maven but I could get both of them in one regex for maven

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 2 above)
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:532]: Fixed.

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 1 above)
Comment (rheh) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:478]: okay

**Review by rheh — APPROVED** [review]
LGMT

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 6]
```diff
@@ -468,10 +470,71 @@ export class WorkspaceTagsService implements IWorkspaceTagsService {
...
+			const artifactId = dependencyContent[2];
+			if (groupId && artifactId) {
+				this.tagJavaDependency(groupId, artifactId, 'workspace.gradle.', tags);
+			}
+		}
+	}
+
+	private tagJavaDependency(groupId: string, artifactId: string, prefix: string, tags: Tags): void {
+		for (const javaLibrary of JavaLibrariesToLookFor) {
+			if ((groupId === javaLibrary.groupId || new RegExp(javaLibrary.groupId).test(groupId)) &&
```
Comment (RMacfarlane) [inline src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:530]: why is this fallback regex needed?

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 7]
```diff
@@ -292,13 +293,13 @@ export class WorkspaceTagsService implements IWorkspaceTagsService {
 			"workspace.py.playwright" : { "classification": "SystemMetaData", "purpose": "FeatureInsight", "isMeasurement": true }
 		}
 	*/
-	private async resolveWorkspaceTags(): Promise<Tags> {
+	private resolveWorkspaceTags(): Promise<Tags> {
 		const tags: Tags = Object.create(null);
```
Comment (sana-ajani) [inline src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: is the [as URI[]] cast necessary?

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts [hunk 8]
```diff
@@ -292,13 +293,13 @@ export class WorkspaceTagsService implements IWorkspaceTagsService {
...
 
 		const state = this.contextService.getWorkbenchState();
```
Comment (RMacfarlane) [inline src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: The `Uri.joinPath` utility would be good to use here

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 6 above)
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts:530]: Though most of the tags could be check by equals directly, some kind of tags are tested by regex, for instance, Azure Java SDK(`azure-mgmt-<service>)`) , test whether `groupId`/`artifactId` equals first for better performance

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 7 above)
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Removed.

File: src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts (hunk 8 above)
Comment (Flanker32) [reply src/vs/workbench/contrib/tags/electron-browser/workspaceTagsService.ts]: Updated~

**Comment by sana-ajani** [comment]
This looks LGTM!

**Pull request merged**